[py] Re-enable remote tests in bazel and fix broken tests

cgoldberg opened the pull request:
### **User description**
### 🔗 Related Issues

### 💥 What does this PR do?

This PR fixes several issue with Python Remote WebDriver tests:
- Fixes bazel `//py:test-remote` target so it picks up tests in `py/test/selenium/webdriver/remote/`. Previously, it was skipping them.
- Moved `pytest_generate_tests` to main `conftest.py`, allowing all tests to access `--driver` flag while removing duplicate `conftest.py` files
- Update the `driver` fixture so it skips Remote WebDriver tests when called with a local driver
- Added new `firefox_options` fixture that returns an instance of `FirefoxOptions` with headless enabled/disabled depending on whether `--headless` arg was given. This is useful in tests that don't use the main `driver` fixture to launch the browser
- Some cleanup in the main `conftest.py`

### 🔄 Types of changes
- Cleanup (formatting, renaming)
- Bug fix (backwards compatible)


___

### **PR Type**
Bug fix, Enhancement


___

### **Description**
- Centralize and fix driver parametrization for all tests
  - Move `pytest_generate_tests` to main `conftest.py`
  - Remove duplicate `conftest.py` files in subdirectories

- Fix and re-enable remote WebDriver tests in Bazel
  - Update Bazel config to include remote test directory
  - Adjust test skipping logic for remote/local driver distinction

- Add new `headless` fixture for consistent headless option access

- Refactor and clean up remote and Firefox test implementations


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>conftest.py</strong><dd><code>Centralize driver parametrization, add headless fixture, improve </code><br><code>skipping logic</code></dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-98e9e290ede5d0c7ac9e7df5b49edf63f5df5fcc946b7736e4a6139e4cda26e3">+22/-4</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>firefox_sizing_tests.py</strong><dd><code>Use new headless fixture and context manager for driver</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-d7416c3031b4161bc84a2aacc443a205a9f9baec189c92fe1376c2d5a683cafd">+8/-15</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>custom_element_tests.py</strong><dd><code>Use fixture for custom element driver, cleanup test logic</code></dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-77a6cf2e863316b3f01fe6c7a26b561e7d04e75b22acf85b868d5b082a2f20ed">+14/-6</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>remote_firefox_profile_tests.py</strong><dd><code>Refactor to use headless fixture and context manager, remove custom </code><br><code>fixtures</code></dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-1939f53bc8bdebf02dbefc8c76c4ad18fb92f8e6d41720058151e9bd3757bde9">+13/-22</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Cleanup</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>conftest.py</strong><dd><code>Remove duplicate driver parametrization logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-4894ee5e83893b4b9e1f2c8014386be4ff9449d33dccf74a236c446df3a09a49">+0/-21</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>conftest.py</strong><dd><code>Remove duplicate driver parametrization logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-da4af917403b65e60b6bbb34c145ace1960b6ecd6fbb949fe9b0f68e1aee70dc">+0/-21</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>conftest.py</strong><dd><code>Remove duplicate driver parametrization logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-8dbcc2f287cbf9d209fae76269f39037d22ab61e2d1e68bd2c447c5d8d41035c">+0/-21</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>conftest.py</strong><dd><code>Remove duplicate driver parametrization logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-e77a168b349808082230e721835a828b7134c0646cdace84e15927b29f4be2d8">+0/-21</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>remote_connection_tests.py</strong><dd><code>Add docstring, clarify test intent for browser-specific method</code></dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-10c82760157398bc5a3ab5ba815d2eb37ed7763ce4999b93379552545d70131d">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>remote_custom_locator_tests.py</strong><dd><code>Mark tests as skipped, add pytest import</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-bd136d071186f266ea0c86c8916238483adac322586a350745e7690f2365fe81">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Formatting</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>remote_downloads_tests.py</strong><dd><code>Minor import reordering</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-076c0af1d631fae433f5422d16b9eccbcac9db99a387833e2662d9bbed250b3a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>remote_hub_connection.py</strong><dd><code>Add options to Remote, improve assertion for SSL error</code>&nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-f0e9d09f74160652aa9aa1086766ce6e300bd1d435a4c242239e9f01f015ae4c">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>BUILD.bazel</strong><dd><code>Update Bazel config to include remote tests, remove obsolete </code><br><code>conftest.py references</code></dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/15657/files#diff-3e8ff3a52e8e6c0b195ea328e17c50ba6d90abca1ce1624110607da20691d7a9">+1/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

cgoldberg committed:
[py] Update bazel config to include tests in remote directory

cgoldberg committed:
[py] Fixed broken remote tests

cgoldberg committed:
[py] Removed unused test markers

titusfortner commented:
Does it need a separate conftest? It'll be easy to overlook it if there are multiple.

cgoldberg committed:
[py] Use fixture

cgoldberg committed:
[py] Fix tests, remove duplicate conftests

cgoldberg committed:
[py] Add headless fixture

cgoldberg commented:
> Does it need a separate conftest?

Good call... it doesn't. I just moved the function it contained into the main conftest, which allowed me to removed it, along with 3 other duplicate conftests.

I just updated the PR description too. This ended up larger than I was expecting, but it cleans up a lot of junk.

I haven't tested it all yet, so we'll see if CI passes :)

cgoldberg commented:
@navin772 

Can you take a look at `py/test/selenium/webdriver/remote/remote_custom_locator_tests.py`? (either in this PR or in trunk once it gets merged).

There are 2 tests that you added in #14587. They fail for me locally, so I have them marked to skip (they weren't getting run in CI before this PR). I don't really understand what they do and don't know if they are failing because of a real issue or they are just broken tests.

cgoldberg committed:
[py] Fix linting error

cgoldberg committed:
[py] Fix options in conftest.py

cgoldberg committed:
[py] Fix typo in fixture name

cgoldberg committed:
[py] Update assertion

cgoldberg committed:
[py] Update remote path detection

cgoldberg committed:
[py] Add missing import

cgoldberg committed:
[py] Fix comment

cgoldberg committed:
[py] Fix build

cgoldberg committed:
[py] Fix remote test identification

navin772 commented:
Hi @cgoldberg, the tests are failing because of a misconfiguration, the remote webdriver expects the `CustomLocatorConverter()` class to be passed as a param for the tests to pass.
Something like this:

```python
driver = webdriver.Remote(command_executor="http://localhost:4444", options=options, locator_converter=CustomLocatorConverter())
```

Do you  know how can I add the above only for my test file?

cgoldberg commented:
@navin772 I'll fix it in this branch. Thanks!

cgoldberg committed:
Merge branch 'trunk' into py-fix-remote-tests

cgoldberg committed:
[py] Fix custom locator tests

cgoldberg committed:
[py] Rename test

cgoldberg commented:
@navin772 thanks.. that worked.

> Do you know how can I add the above only for my test file?

I added a small fixture to the test file so it doesn't use the main `driver` fixture. It's not great, but it solves this problem.

navin772 commented:
@cgoldberg thanks for the changes, this PR looks good, just run the formatter once.

cgoldberg committed:
[py] Fix linting error

cgoldberg committed:
Merge branch 'SeleniumHQ:trunk' into py-fix-remote-tests

titusfortner commented:
>  which allowed me to removed it, along with 3 other duplicate conftests

Right, I had no idea there were multiple of these, so thanks!

titusfortner commented on py/test/selenium/webdriver/firefox/firefox_sizing_tests.py:
```diff
@@ -30,33 +30,26 @@ def is_running_wayland():
-    options = Options()
-    if request.config.getoption("--headless"):
-        options.add_argument("-headless")
+def test_firefox_opens_large_when_running_xwayland(headless):
```
having a conditional in these tests isn't a good practice. Can we put the conditional in the fixture (maybe call it firefox_options) which adds headless arg if it applies?

cgoldberg replied:
I did that in a few tests.. I have no idea why :) I'll move it to a fixture.

cgoldberg committed:
[py] Move headless option to fixture

cgoldberg committed:
Merge branch 'SeleniumHQ:trunk' into py-fix-remote-tests

cgoldberg commented:
The only failure in CI is not related to this PR.
